Add changes for graceful node decommission

pranikum opened the pull request:
Signed-off-by: pranikum <109206473+pranikum@users.noreply.github.com>

### Description
Changes for graceful decommission of nodes. This includes call to WRR api.
Can be merged once decommission API and WRR changes are merged.

Depends on :
https://github.com/opensearch-project/OpenSearch/pull/4272
https://github.com/opensearch-project/OpenSearch/pull/4241

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/4083

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

gbbafna commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +277,154 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        String reason,
+        TimeValue timeout,
+        ActionListener<Void> nodesRemovedListener,
+        TimeValue timeoutForNodeDecommission
```
nit : `timeoutForNodeDraining` rather than timeoutForNodeDecommission ? Decommission is whole process , of which node removal is part .

pranikum replied:
Done.

pranikum commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +277,154 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        checkHttpStatsForDecommissionedNodes(nodesToBeDecommissioned, reason, timeout, timeoutForNodeDecommission, nodesRemovedListener);
+    }
+
+    void setWeightForDecommissionedZone(List<String> zones) {
```
We are doing WRR for graceful decommission. On successful weighing away of the zone we will remove the nodes from the decommissioned zone after the timeout has expired.

pranikum commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +277,154 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+
+        // WRR API will validate invalid weights
+        final ClusterPutWRRWeightsRequest clusterWeightRequest = new ClusterPutWRRWeightsRequest();
+        clusterWeightRequest.attributeName("zone");
```
decommissionAttribute.attributeName it should be. Updated the same.

pranikum commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +277,154 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+
+        if (timeoutForNodeDecommission.getSeconds() > 0) {
+            // Wait for timeout to happen. Log the active connection before decommissioning of nodes.
+            scheduleDecommissionNodesRequestCheck(decommissionedNodes, reason, timeout, listener, timeoutForNodeDecommission);
```
Yes... It will get scheduled immediately. So else block is not needed.

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +277,128 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+    void setWeightForDecommissionedZone(List<String> zones, ActionListener<ClusterPutWRRWeightsResponse> listener) {
+        ClusterState clusterState = clusterService.getClusterApplierService().state();
+
+        DecommissionAttributeMetadata decommissionAttributeMetadata = clusterState.metadata().custom(DecommissionAttributeMetadata.TYPE);
```
Use utility function created in metadata for getting decommission attribute - `clusterState.metadata().decommissionedAttribute()`

pranikum replied:
Done. Used clusterState.metadata().decommissionedAttributeMetadata().

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +277,128 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
+
+        Map<String, String> weights = new HashMap<>();
+        zones.forEach(zone -> {
```
avoid keyword `zone`

pranikum replied:
Done. Changed to awarenessValue.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +277,128 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        Map<String, String> weights = new HashMap<>();
+        zones.forEach(zone -> {
+            if (zone.equalsIgnoreCase(decommissionAttribute.attributeValue())) {
+                weights.put(zone, "0");
```
is the weight string or float?

pranikum replied:
Its float. Taken the latest Changes from WRR API PR. Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +277,128 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+                public void handleException(TransportException exp) {
+                    // Logging warn message on failure. Should we do Retry? If weights are not set should we fail?
+                    logger.warn("Exception occurred while setting weights.Exception Messages - [{}]", exp.unwrapCause().getMessage());
+                    listener.onFailure(exp);
```
How are we handling failures?

pranikum replied:
The listener will call clearVotingConfig with appropriate parameter

```
                decommissionController.setWeight(awarenessValues, new ActionListener<>() {
                    @Override
                    public void onResponse(ClusterPutWeightedRoutingResponse response) {
                        scheduleNodesDecommissionOnTimeout(state, timeOutForNodeDraining);
                    }

                    @Override
                    public void onFailure(Exception e) {
                        clearVotingConfigExclusionAndUpdateStatus(false, false);
                    }
                });
```

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +277,128 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        TimeValue timeoutForNodeDraining,
+        ActionListener<Void> listener
+    ) {
+        // Wait for timeout to happen. Log the active connection before decommissioning of nodes.
+        transportService.getThreadPool().schedule(new Runnable() {
+            @Override
+            public void run() {
+                // Check for active connections.
+                getActiveRequestCountOnDecommissionNodes(decommissionedNodes);
+                removeDecommissionedNodes(decommissionedNodes, reason, timeout, listener);
+            }
+        }, timeoutForNodeDraining, org.opensearch.threadpool.ThreadPool.Names.SAME);
+    }
```
Didn't get this. `removeDecommissionedNodes` should only be called once after node is drained

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -117,7 +118,8 @@ private void setForcedAwarenessAttributes(Settings forceSettings) {
     public void startDecommissionAction(
         final DecommissionAttribute decommissionAttribute,
-        final ActionListener<ClusterStateUpdateResponse> listener
+        final ActionListener<ClusterStateUpdateResponse> listener,
+        final TimeValue timeOutForNodeDraining
```
nit: rename to `drainingTimeout`

imRishN replied:
Accept the entire request object here so that we don't always need to manually update the method here and just modifying the object will serve the purpose

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -156,14 +158,15 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
                     decommissionAttributeMetadata.status()
                 );
-                decommissionClusterManagerNodes(decommissionAttributeMetadata.decommissionAttribute(), listener);
+                decommissionClusterManagerNodes(decommissionAttributeMetadata.decommissionAttribute(), listener, timeOutForNodeDraining);
```
Same here. Send the entire request to the method

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -156,14 +158,15 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         final DecommissionAttribute decommissionAttribute,
-        ActionListener<ClusterStateUpdateResponse> listener
+        ActionListener<ClusterStateUpdateResponse> listener,
+        TimeValue timeOutForNodeDraining
```
nit: rename

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -206,7 +209,10 @@ public void onResponse(Void unused) {
                         listener.onResponse(new ClusterStateUpdateResponse(true));
-                        failDecommissionedNodes(clusterService.getClusterApplierService().state());
+
+                        weighAwayDecommissionedZone(state, timeOutForNodeDraining);
```
nit: lets call it `drainNodesWithDecommissionedAttribute`.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -303,19 +309,66 @@ public void onFailure(Exception e) {
+        // Awareness values refers to all zones in the cluster
+        List<String> awarenessValues = forcedAwarenessAttributes.get(decommissionAttribute.attributeName());
+
+        decommissionController.updateMetadataWithDecommissionStatus(DecommissionStatus.WEIGH_AWAY, new ActionListener<>() {
```
nit: We need a verb as status. Can we use `DRAINING`?

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -303,19 +309,66 @@ public void onFailure(Exception e) {
+                decommissionController.setWeightForDecommissionedZone(awarenessValues, new ActionListener<>() {
+                    @Override
+                    public void onResponse(ClusterPutWRRWeightsResponse response) {
+                        clearVotingConfigExclusionAndUpdateStatus(true, true);
```
Why are we clearing the exclusion here and updating status?

pranikum replied:
Fixed. Thanks for pointing out.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -303,19 +309,66 @@ public void onFailure(Exception e) {
+
+                decommissionController.setWeightForDecommissionedZone(awarenessValues, new ActionListener<>() {
+                    @Override
+                    public void onResponse(ClusterPutWRRWeightsResponse response) {
```
Does this response just sets the weight or also ensure node is drained?

pranikum replied:
It just set the weight. We are adding timeout for nodes to be drained and log the active connection on the nodes being decommissioned.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -303,19 +309,66 @@ public void onFailure(Exception e) {
                     filterNodesWithDecommissionAttribute(clusterService.getClusterApplierService().state(), decommissionAttribute, false),
                     "nodes-decommissioned",
                     TimeValue.timeValueSeconds(120L),
+                    timeOutForNodeDraining,
```
node removal method should be independent of draining

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionStatus.java:
```diff
@@ -16,6 +16,11 @@ public enum DecommissionStatus {
      * Decommission process is initiated, and to-be-decommissioned leader is excluded from voting config
      */
     INIT("init"),
+    /**
+     * Decommission process is initiated, and the zone is being weighed away.
+     */
+    WEIGH_AWAY("weigh_away"),
+
```
The life cycle of this new status should be - 
1. Setting the weights
2. Wait for it to drain
3. Handle failures of above two gracefully
4. Only on success of first two we should call node removal as is

We can rename this to `DRAINING` or `WEIGHING_AWAY`. I prefer first

pranikum replied:
Updated the name to `DRAINING`

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionStatus.java:
```diff
@@ -56,6 +61,8 @@ public static DecommissionStatus fromString(String status) {
         }
         if (status.equals(INIT.status())) {
             return INIT;
+        } else if (status.equals(WEIGH_AWAY.status())) {
+            return WEIGH_AWAY;
         } else if (status.equals(IN_PROGRESS.status())) {
             return IN_PROGRESS;
         } else if (status.equals(SUCCESSFUL.status())) {
```
Will come back to state handling in next round of review

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,94 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+
+    void setWeights(String awarenessAttribute, Map<String, Double> weights, ActionListener<ClusterPutWeightedRoutingResponse> listener) {
+        // WRR API will validate invalid weights
+        final ClusterPutWeightedRoutingRequest clusterWeightRequest = new ClusterPutWeightedRoutingRequest();
```
nit: clusterWeightRequest  -> clusterPutRoutingWeightRequest

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,94 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         });
     }
+
+    void setWeights(String awarenessAttribute, Map<String, Double> weights, ActionListener<ClusterPutWeightedRoutingResponse> listener) {
```
Is awarenessAttribute a key to the attribute? For eg. zone? If yes, then we need to reword this as attribute is key-value pair and not just key of the attribute

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,94 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+            new TransportResponseHandler<ClusterPutWeightedRoutingResponse>() {
+                @Override
+                public void handleResponse(ClusterPutWeightedRoutingResponse response) {
+                    logger.info("Weights are successfully set.");
```
Let's log the response and update the log msg

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -206,7 +210,10 @@ public void onResponse(Void unused) {
                         listener.onResponse(new ClusterStateUpdateResponse(true));
-                        failDecommissionedNodes(clusterService.getClusterApplierService().state());
+
+                        weighAwayDecommissionedZone(state, decommissionRequest);
```
Since the status is `DRAINING` lets call this method `drainNodesWithDecommissionedAttribute`

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -303,15 +310,98 @@ public void onFailure(Exception e) {
+        });
+    }
+
+    void setWeightsForAwarenessAttribute(List<String> awarenessValues, ClusterState state, DecommissionRequest decommissionRequest) {
```
nit: rename method to `setRoutingWeightsToAwarenessAttribute`

pranikum replied:
Renamed to ```setRoutingWeights```

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -303,15 +310,98 @@ public void onFailure(Exception e) {
+        });
+    }
+
+    void setWeightsForAwarenessAttribute(List<String> awarenessValues, ClusterState state, DecommissionRequest decommissionRequest) {
```
Do we need state here in this method? As we are getting it from applier service?

pranikum replied:
Removed the param to method.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -303,15 +310,98 @@ public void onFailure(Exception e) {
+        });
+    }
+
+    void scheduleNodesDecommissionOnTimeout(ClusterState state, TimeValue timeoutForNodeDraining) {
```
get state from applier service in this method

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java:
```diff
@@ -0,0 +1,110 @@
+        if (decommissionAttribute.attributeValue() == null || decommissionAttribute.attributeValue().isEmpty()) {
+            validationException = addValidationError("attribute value is missing", validationException);
+        }
+        if (drainingTimeout == null || drainingTimeout.getSeconds() < 0) {
+            validationException = addValidationError("Invalid draining timeout", validationException);
+        }
```
Ignore this comment as explicit transport call in future might not be setting it

pranikum replied:
Have added default timeout value in the request class itself.

imRishN commented:
Should we also have a force flag in the request such that if the WeightedRouting transport call fails, the request will still honour the decommission request and kick the decommissioned node out of the cluster?

pranikum commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,94 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         });
     }
+
+    void setWeights(String awarenessAttribute, Map<String, Double> weights, ActionListener<ClusterPutWeightedRoutingResponse> listener) {
```
Let me rethink of a name. We will just set the weights in this method. This method will be used for recommission as well. And we can pass the weights and keep the call to weights api generic. Hence i have split the methods to avoid code duplication.

pranikum replied:
Updated to ```setRoutingWeights```

pranikum commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -303,15 +310,98 @@ public void onFailure(Exception e) {
+            @Override
+            public void onResponse(ClusterPutWeightedRoutingResponse response) {
+                // Schedule the node decommission process after the weights are successfully set.
+                scheduleNodesDecommissionOnTimeout(state, decommissionRequest.getDrainingTimeout());
```
Taken for applier service

pranikum commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,94 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        final ClusterPutWeightedRoutingRequest clusterWeightRequest = new ClusterPutWeightedRoutingRequest();
+        clusterWeightRequest.attributeName(awarenessAttribute);
+        clusterWeightRequest.setWeightedRouting(new WeightedRouting(awarenessAttribute, weights));
+
```
As discussed. It will fail during the transport call. So it should be ok.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,99 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+
+    void setRoutingWeights(
+        String awarenessAttributeName,
+        Map<String, Double> weights,
```
Lets call this awarenessAttributeToWeightMap?

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,99 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+                @Override
+                public void handleException(TransportException exp) {
+                    // Logging warn message on failure. Should we do Retry? If weights are not set should we fail?
+                    logger.error("Exception occurred while setting weights.Exception Messages - [{}]", exp.unwrapCause().getMessage());
```
Lets add the entire exp

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,99 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        logger.info("Decommissioning node with connections : [{}]", nodeActiveConnectionMap);
+    }
+
+    void getActiveRequestCountOnDecommissionNodes(Set<DiscoveryNode> decommissionedNodes) {
```
nit: decommissionedNodes

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,99 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+            new TransportResponseHandler<NodesStatsResponse>() {
+                @Override
+                public void handleResponse(NodesStatsResponse response) {
+                    logActiveConnections(response);
```
Can response be null?

pranikum replied:
Handled for null response. Added log for null response.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,99 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+
+                @Override
+                public void handleException(TransportException exp) {
+                    logger.warn("Failure occurred while dumping connection for decommission nodes - [{}]", exp.unwrapCause().getMessage());
```
Same as above. Put the entire exp object

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -303,15 +310,95 @@ public void onFailure(Exception e) {
         DecommissionAttributeMetadata decommissionAttributeMetadata = state.metadata().decommissionAttributeMetadata();
         DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
+
+        // Awareness values refers to all zones in the cluster
+        List<String> awarenessValues = forcedAwarenessAttributes.get(decommissionAttribute.attributeName());
```
Lets not add that comment. `zone` is unknown to OpenSearch. Also rename this to `attributeValues`

pranikum replied:
Removed the comment

imRishN commented on server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java:
```diff
@@ -28,23 +29,36 @@
 
     public DecommissionRequest(DecommissionAttribute decommissionAttribute) {
         this.decommissionAttribute = decommissionAttribute;
+        //
```
nit: remove

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestDecommissionAction.java:
```diff
@@ -46,9 +46,19 @@ protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient cli
+        String attributeName = null;
+        String attributeValue = null;
         DecommissionRequest decommissionRequest = Requests.decommissionRequest();
-        String attributeName = request.param("awareness_attribute_name");
-        String attributeValue = request.param("awareness_attribute_value");
-        return decommissionRequest.setDecommissionAttribute(new DecommissionAttribute(attributeName, attributeValue));
+        if (request.hasParam("awareness_attribute_name")) {
+            attributeName = request.param("awareness_attribute_name");
+        }
+
+        if (request.hasParam("awareness_attribute_value")) {
+            attributeValue = request.param("awareness_attribute_value");
+        }
+        decommissionRequest.setDecommissionAttribute(new DecommissionAttribute(attributeName, attributeValue));
+        return decommissionRequest.setDrainingTimeout(
+            request.paramAsTime("draining_timeout", DecommissionRequest.DEFAULT_NODE_DRAINING_TIMEOUT)
+        );
```
Please revert the unexpected changes

pranikum replied:
Done.

pranikum commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -207,7 +211,8 @@ public void onResponse(Void unused) {
                         listener.onResponse(new DecommissionResponse(true));
-                        failDecommissionedNodes(clusterService.getClusterApplierService().state());
+
+                        drainNodesWithDecommissionedAttribute(state, decommissionRequest);
```
Taken from clusterApplierService.

pranikum commented on server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java:
```diff
@@ -74,6 +103,9 @@ public ActionRequestValidationException validate() {
             validationException = addValidationError("attribute value is missing", validationException);
         }
+        if (drainingTimeout.getSeconds() < 0) {
+            validationException = addValidationError("Invalid draining timeout", validationException);
```
This value comes from request. value < 0 is not expected hence I feel invalid makes more sense. Let me know if you have more appropriate suggestion.

imRishN replied:
Lets the add the timeout as well in the message

imRishN commented on server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java:
```diff
@@ -28,23 +29,35 @@
         super(in);
         decommissionAttribute = new DecommissionAttribute(in);
+        this.drainingTimeout = in.readTimeValue();
+
```
nit: remove extra line

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,104 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        Map<String, Double> awarenessAttributeToWeightMap,
+        ActionListener<ClusterPutWeightedRoutingResponse> listener
+    ) {
+        // WRR API will validate invalid weights
```
nit: WRR transport action

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,104 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+            new TransportResponseHandler<ClusterPutWeightedRoutingResponse>() {
+                @Override
+                public void handleResponse(ClusterPutWeightedRoutingResponse response) {
+                    logger.info("Weights are successfully set. [{}]", clusterPutRoutingWeightRequest.getWeightedRouting().weights());
```
Do we need to use toString?

pranikum replied:
Not needed. It will be called implicitly

Bukhtawar replied:
Can we directly use `WeightedRoutingService#registerWeightedRoutingMetadata` instead

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,104 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+                @Override
+                public void handleException(TransportException exp) {
+                    // Logging message on failure. Should we do Retry? If weights are not set should we fail?
+                    logger.error("Exception occurred while setting weights.Exception Messages - ", exp);
```
`.Exception Messages` looks a little weird here
Maybe we can add the message like - 
`Failed to set routing weights [{}], weights and pass the exception

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -112,13 +114,14 @@ private void setForcedAwarenessAttributes(Settings forceSettings) {
+        final DecommissionRequest decommissionRequest,
         final ActionListener<DecommissionResponse> listener
     ) {
+        DecommissionAttribute decommissionAttribute = decommissionRequest.getDecommissionAttribute();
```
Let's make this final?

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -304,10 +309,88 @@ public void onFailure(Exception e) {
 
-    private void failDecommissionedNodes(ClusterState state) {
+    private void drainNodesWithDecommissionedAttribute(ClusterState state, DecommissionRequest decommissionRequest) {
+        // this method ensures no matter what, we always exit from this function after clearing the voting config exclusion
+        DecommissionAttributeMetadata decommissionAttributeMetadata = state.metadata().decommissionAttributeMetadata();
+        DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
+
+        List<String> awarenessValues = forcedAwarenessAttributes.get(decommissionAttribute.attributeName());
+
```
this piece of code is not needed as you are anyways get the latest state from applier again and can use the `awarenessValues` from that itself

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -304,10 +309,88 @@ public void onFailure(Exception e) {
+    void setRoutingWeightsToAwarenessAttribute(List<String> awarenessAttributeValues, DecommissionRequest decommissionRequest) {
+        ClusterState state = clusterService.getClusterApplierService().state();
+
+        DecommissionAttributeMetadata decommissionAttributeMetadata = state.metadata().decommissionAttributeMetadata();
+        assert decommissionAttributeMetadata.status().equals(DecommissionStatus.DRAINING)
+            : "unexpected status encountered while decommissioning nodes";
+        DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
```
What if the metadata was wiped off by the time we come here? Shall we keep a null check and call ` clearVotingConfigExclusionAndUpdateStatus(false, false);` if null found

pranikum replied:
Done.

imRishN reviewed (approved):
Thanks @pranikum for the changes

imRishN commented:
Hey, i think we missed one case as part of this PR. If node is in `DRAINING` state, we should block the joins. Today we are doing it for IN_PROGRESS and SUCCESSFUL. Let's add draining as well here. One such place where we need to add the check is - 
```
public static boolean nodeCommissioned(DiscoveryNode discoveryNode, Metadata metadata) {
        DecommissionAttributeMetadata decommissionAttributeMetadata = metadata.decommissionAttributeMetadata();
        if (decommissionAttributeMetadata != null) {
            DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
            DecommissionStatus status = decommissionAttributeMetadata.status();
            if (decommissionAttribute != null && status != null) {
                if (nodeHasDecommissionedAttribute(discoveryNode, decommissionAttribute)
                    && (status.equals(DecommissionStatus.IN_PROGRESS) || status.equals(DecommissionStatus.SUCCESSFUL))) {
                    return false;
                }
            }
        }
        return true;
    }
```

imRishN reviewed (changes requested):
Hey, i think we missed one case as part of this PR. If node is in DRAINING state, we should block the joins. Today we are doing it for IN_PROGRESS and SUCCESSFUL. Let's add draining as well here. One such place where we need to add the check is -

```
public static boolean nodeCommissioned(DiscoveryNode discoveryNode, Metadata metadata) {
        DecommissionAttributeMetadata decommissionAttributeMetadata = metadata.decommissionAttributeMetadata();
        if (decommissionAttributeMetadata != null) {
            DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
            DecommissionStatus status = decommissionAttributeMetadata.status();
            if (decommissionAttribute != null && status != null) {
                if (nodeHasDecommissionedAttribute(discoveryNode, decommissionAttribute)
                    && (status.equals(DecommissionStatus.IN_PROGRESS) || status.equals(DecommissionStatus.SUCCESSFUL))) {
                    return false;
                }
            }
        }
        return true;
    }
```

pranikum commented:
> Hey, i think we missed one case as part of this PR. If node is in DRAINING state, we should block the joins. Today we are doing it for IN_PROGRESS and SUCCESSFUL. Let's add draining as well here. One such place where we need to add the check is -
> 
> ```
> public static boolean nodeCommissioned(DiscoveryNode discoveryNode, Metadata metadata) {
>         DecommissionAttributeMetadata decommissionAttributeMetadata = metadata.decommissionAttributeMetadata();
>         if (decommissionAttributeMetadata != null) {
>             DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
>             DecommissionStatus status = decommissionAttributeMetadata.status();
>             if (decommissionAttribute != null && status != null) {
>                 if (nodeHasDecommissionedAttribute(discoveryNode, decommissionAttribute)
>                     && (status.equals(DecommissionStatus.IN_PROGRESS) || status.equals(DecommissionStatus.SUCCESSFUL))) {
>                     return false;
>                 }
>             }
>         }
>         return true;
>     }
> ```

Sure will do... Thanks

imRishN commented:
Do you think we need a force flag here in this case? Let's say user keep the draining timeout to 7 min. 7 min is a huge time and can have master switches in between. If at all there's a master switch where the thread was doing nothing but sleeping, in that case, the decommission request would just get abandoned, which is fine. But IMO, we need two things here - 
1. A force flag where the user can just skip the sleep
2. Request count check on to-be-decommissioned node just after setting the weights. And if it is already zero, just skip the sleep and proceed

pranikum commented:
> Do you think we need a force flag here in this case? Let's say user keep the draining timeout to 7 min. 7 min is a huge time and can have master switches in between. If at all there's a master switch where the thread was doing nothing but sleeping, in that case, the decommission request would just get abandoned, which is fine. But IMO, we need two things here -
> 
> 1. A force flag where the user can just skip the sleep, so that in the next request he again doesn't need to wait for draining timeout unnecessarily
> 2. Request count check on to-be-decommissioned node just after setting the weights. And if it is already zero, just skip the sleep and proceed

Thats why we have node drain timeout.. If that is set to zero it will skip the sleep. The  scheduler framework is already handling that. In that case we just set the weight log the active requests and decommission the nodes. I feel that what we need. Do yo see any specific need of force flag.

pranikum commented:
> > Do you think we need a force flag here in this case? Let's say user keep the draining timeout to 7 min. 7 min is a huge time and can have master switches in between. If at all there's a master switch where the thread was doing nothing but sleeping, in that case, the decommission request would just get abandoned, which is fine. But IMO, we need two things here -
> > 
> > 1. A force flag where the user can just skip the sleep, so that in the next request he again doesn't need to wait for draining timeout unnecessarily
> > 2. Request count check on to-be-decommissioned node just after setting the weights. And if it is already zero, just skip the sleep and proceed
> 
> Thats why we have node drain timeout.. If that is set to zero in the request it will skip the sleep. The scheduler framework is already handling that. In that case we just set the weight, log the active requests and decommission the nodes. I feel that's what we need. Do yo see any specific need of force flag.

Created 2 separate tickets to discuss on these items.
https://github.com/opensearch-project/OpenSearch/issues/4721
https://github.com/opensearch-project/OpenSearch/issues/4720

imRishN commented:
Also, can it happen that during the first attempt of setting weights to decommissioned attribute, it was already weighed out by different mechanisms the user might already be having and hence by the time the user decommissioned the attribute - 
1. It was already weighed away
2. It was possibly already drained

In that case, the current code flow would be doing work which was already done in the past before executing decommission. For a cluster under stress, it is paramount for it to not accept such works which was already done. 

@pranikum let me know your thoughts on this

imRishN commented:
Putting the above discussion on gracefully handling the decommission below - 

Do you think we need a force flag here in this case? Let's say user keep the draining timeout to 7 min. 7 min is a huge time and can have master switches in between. If at all there's a master switch where the thread was doing nothing but sleeping, in that case, the decommission request would just get abandoned, which is fine. But IMO, we need two things here -

1. A force flag where the user can just skip the sleep, so that in the next request he again doesn't need to wait for draining timeout unnecessarily
2. Request count check on to-be-decommissioned node just after setting the weights. And if it is already zero, just skip the sleep and proceed


Can it happen that during the first attempt of setting weights to decommissioned attribute, it was already weighed out by different mechanisms the user might already be having and hence by the time the user decommissioned the attribute -

1. It was already weighed away
2. It was possibly already drained

In that case, the current code flow would be doing work which was already done in the past before executing decommission. For a cluster under stress, it is paramount for it to not accept such works which was already done.

Bukhtawar commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,104 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        clusterPutRoutingWeightRequest.setWeightedRouting(new WeightedRouting(awarenessAttributeName, awarenessAttributeToWeightMap));
+
+        transportService.sendRequest(
+            transportService.getLocalNode(),
```
Should we assert local node is cluster manager node?

pranikum replied:
Done.

Bukhtawar commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -305,10 +309,87 @@ public void onFailure(Exception e) {
+            decommissionController.getActiveRequestCountOnDecommissionedNodes(decommissionedNodes);
+            // Call to fail the decommission nodes
+            failDecommissionedNodes(clusterService.getClusterApplierService().state(), decommissionAttribute);
+        }, timeoutForNodeDraining, org.opensearch.threadpool.ThreadPool.Names.SAME);
```
Please use a GENERIC thread pool

pranikum replied:
Done.

Bukhtawar commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -305,10 +309,87 @@ public void onFailure(Exception e) {
+        List<String> awarenessAttributeValues = forcedAwarenessAttributes.get(decommissionAttribute.attributeName());
+
+        Map<String, Double> weights = new HashMap<>();
+        awarenessAttributeValues.forEach(awarenessValue -> {
+            if (awarenessValue.equalsIgnoreCase(decommissionAttribute.attributeValue())) {
+                weights.put(awarenessValue, Double.valueOf(0.0));
+            } else {
+                weights.put(awarenessValue, Double.valueOf(1.0));
+            }
+        });
```
what if the weights were 3:2:1 to start with would that be reset back?

pranikum replied:
@Bukhtawar .. Currently we are not setting the weights back to original value. We are leaving currently it to same value and rely the client to set the appropriate weights back. 
I am opening a task on discussing the same. 

https://github.com/opensearch-project/OpenSearch/issues/4803

Bukhtawar replied:
why can't we do 3:2:0 in the above example

Bukhtawar replied:
Can we do 3:2:0 instead of 1:1:0?

pranikum replied:
@Bukhtawar .. Not sure if i have understood this.. Are you saying to set different weights for the attributes? And how to decide which attribute should have 3 or 2. This can lead to unbalanced traffic movement.

imRishN replied:
@pranikum Consider this example -
Let's say before decommission the weights were z1 -> 3, z2 -> 2, z3 -> 1

Now as per the current implementation, when we decommission z3, we would set the weights to z1 -> 1, z2 -> 1, z3 -> 0. @Bukhtawar's point is rather than doing this, only set the decommissioned zone weight to 0 and leave other zone's weight untouched. Something like, z1-> 3, z2 -> 2, z1 -> 0

But this again has a caveat to it when a different non decommissioned zone already had weight 0 (z1 -> 0, z2 -> 2, z3 -> 1) and if we decommission z3 and set weight to (z1 -> 0, z2 -> 2, z3 -> 0) this again can be problem.

Now, we don't know why the user has set weight for z1 as 0 but again we cannot let weights of 2 zones to be 0.

pranikum replied:
Ok.. I understood the situation now. But this can lead to 2 zone with weights 0. If we go by this flow as well.
Initial weights : z1 -> 3 z2 -> 2 z3 -> 1
Now lets say we have these actions 
1. Decommission z3 - (Weights after this step z1 : 3 z2: 2 z3: 0) 
2. Recommission z3 - (Weights after this step z1 : 3 z2: 2 z3: 0) (Since currently we are not setting weights back) 
3. Decommission z1 - (Weights after this step z1 : 0 z2: 2 z3: 0)  Here we have 2 decommissioned zone.
4. Recommission z1 - (Weights after this step z1 : 0 z2: 2 z3: 0) 

I feel we should do this after we have done https://github.com/opensearch-project/OpenSearch/issues/4803
In this case weights will be set back during recommission and will not lead to multiple weights being 0.

imRishN replied:
For this case, I think we should let the weights of not decommissioned zone as is. If a non decommissioned zone had weight zero then we can assign the weight of decommissioned zone to it, and put weight of decommissioned zone to zero
@Bukhtawar thoughts on this?

Bukhtawar replied:
Should we have a mechanism to set weights as -1 for decommission to distinguish from an explicit weight as 0. That way we reject the second decommission call if weights are unspecified(-1)

pranikum replied:
As disscussed. We will not be setting weights for now. We will let the user to set the weights.

Bukhtawar reviewed (changes requested):
Thanks @pranikum left few comments

pranikum commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,104 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+            return;
+        }
+
+        Map<String, Long> nodeActiveConnectionMap = new HashMap<>();
+        List<NodeStats> responseNodes = nodesStatsResponse.getNodes();
+        for (int i = 0; i < responseNodes.size(); i++) {
+            HttpStats httpStats = responseNodes.get(i).getHttp();
+            DiscoveryNode node = responseNodes.get(i).getNode();
+            nodeActiveConnectionMap.put(node.getId(), httpStats.getServerOpen());
+        }
+        logger.info("Decommissioning node with connections : [{}]", nodeActiveConnectionMap);
```
@Bukhtawar .. We are just logging the active connections for now. I was thinking once we have datapoint related to active connections count for the nodes we can set the minimum connection setting. I will open a task to track that change. Let me know your thoughts.

pranikum commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -267,4 +278,104 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        Map<String, Double> awarenessAttributeToWeightMap,
+        ActionListener<ClusterPutWeightedRoutingResponse> listener
+    ) {
+        // WRR transport action will validate invalid weights
```
Done

pranikum commented on server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java:
```diff
@@ -58,13 +70,28 @@ public DecommissionRequest setDecommissionAttribute(DecommissionAttribute decomm
+     */
+    public DecommissionRequest setDrainingTimeout(TimeValue timeout) {
+        this.drainingTimeout = timeout;
+        return this;
```
Following the other methods pattern. Other classes have the same pattern. Like `ClusterHealthRequest`

pranikum commented on server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java:
```diff
@@ -74,6 +101,9 @@ public ActionRequestValidationException validate() {
             validationException = addValidationError("attribute value is missing", validationException);
         }
+        if (drainingTimeout.getSeconds() < 0) {
+            validationException = addValidationError("Invalid draining timeout - " + drainingTimeout.getSeconds(), validationException);
```
Have Updated. Please check if thats fine.

pranikum commented on server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java:
```diff
@@ -28,23 +29,34 @@
+        this.decommissionAttribute = decommissionAttribute;
+        this.drainingTimeout = drainingTimeout;
+    }
+
```
Done.

imRishN commented:
This PR missed adding validation in `DecommissionedAttributeMetadata`. The status update to `DRAINING` should have failed if not done. Did we see status getting updated to `DRAINING`?

pranikum commented:
> This PR missed adding validation in `DecommissionedAttributeMetadata`. The status update to `DRAINING` should have failed if not done. Did we see status getting updated to `DRAINING`? @pranikum Let's add integ test to check that **ALL** the nodes has draining status when the leader state says `DRAINING` in the metadata

Thanks for mentioning that. I have added the validation for state transition. Also added Unit tests to capture all state transition. Will add this in integration test as well

pranikum commented:
@Bukhtawar .. Regarding the comment
```
Can we directly use WeightedRoutingService#registerWeightedRoutingMetadata instead
```

We can use that as well. But the issue will be that we will be missing the validation check on the input done in the action class. Also if any further logic is added down the line it will not be executed in this flow

imRishN commented on server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java:
```diff
@@ -28,23 +29,36 @@
  */
 public class DecommissionRequest extends ClusterManagerNodeRequest<DecommissionRequest> {
 
+    public static final TimeValue DEFAULT_NODE_DRAINING_TIMEOUT = TimeValue.timeValueSeconds(300);
```
Let's keep default to 2 min? As the node was already weighed away and 5 min can be too large a number just for inflight requests?

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -305,10 +309,73 @@ public void onFailure(Exception e) {
+
+    private void ensureDecommissionedAttributeWeightsAreSet(ClusterState state, DecommissionAttribute decommissionAttribute) {
+        WeightedRoutingMetadata weightedRoutingMetadata = state.metadata().weightedRoutingMetadata();
+        assert weightedRoutingMetadata != null && weightedRoutingMetadata.getWeightedRouting() != null;
+        assert weightedRoutingMetadata.getWeightedRouting().weights().get(decommissionAttribute.attributeValue()) == 0.0;
```
this won't run in prod environment. Also remove this check entirely. We are handling it as part of #4839

pranikum replied:
Removed.

Bukhtawar commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -271,4 +278,62 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        if (decommissionedNodes == null || decommissionedNodes.isEmpty()) {
+            return;
+        }
+
```
nit: avoid extra line breaks

pranikum replied:
Done.

Bukhtawar commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -271,4 +278,62 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+
+                @Override
+                public void handleException(TransportException exp) {
+                    logger.warn("Failure occurred while dumping connection for decommission nodes - [{}]", exp.unwrapCause().getMessage());
```
Maybe log the exception with the full trace

pranikum replied:
Done.

Bukhtawar commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -305,10 +309,73 @@ public void onFailure(Exception e) {
+            : "Unexpected status encountered while decommissioning nodes.";
+        DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
+
+        ensureDecommissionedAttributeWeightsAreSet(state, decommissionAttribute);
```
Lets move this check in the cluster state call prior to updating the state to `DRAINING`. Technically its a pre-req for draining to even start

pranikum replied:
We have removed this check since it is being done in https://github.com/opensearch-project/OpenSearch/pull/4839

Bukhtawar commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -305,10 +309,73 @@ public void onFailure(Exception e) {
+
+        Set<DiscoveryNode> decommissionedNodes = filterNodesWithDecommissionAttribute(state, decommissionAttribute, false);
+        // Wait for timeout to happen. Log the active connection before decommissioning of nodes.
+        transportService.getThreadPool().schedule(() -> {
```
This should be `GENERIC` thread pool

pranikum replied:
Done.

Bukhtawar commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -305,10 +309,73 @@ public void onFailure(Exception e) {
+    void checkWeightsAndScheduleDecommission(DecommissionRequest decommissionRequest) {
+        ClusterState state = clusterService.getClusterApplierService().state();
+
+        DecommissionAttributeMetadata decommissionAttributeMetadata = state.metadata().decommissionAttributeMetadata();
+        if (decommissionAttributeMetadata == null) {
+            clearVotingConfigExclusionAndUpdateStatus(false, false);
+            return;
+        }
+        assert decommissionAttributeMetadata.status().equals(DecommissionStatus.DRAINING)
+            : "Unexpected status encountered while decommissioning nodes.";
+        DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
```
When after the `DRAINING` begins can metadata be null?

pranikum replied:
I think yes.. Delete decommission API can be called which can clear the metadata

imRishN commented on server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java:
```diff
@@ -74,6 +103,12 @@ public ActionRequestValidationException validate() {
         if (decommissionAttribute.attributeValue() == null || Strings.isEmpty(decommissionAttribute.attributeValue())) {
             validationException = addValidationError("attribute value is missing", validationException);
         }
+        if (drainingTimeout.getSeconds() < 0 || drainingTimeout.getSeconds() > MAX_NODE_DRAINING_TIMEOUT.getSeconds()) {
```
get millis?

pranikum replied:
Using getSeconds only. Don't see any specific benefit and will make it confusing since we are using seconds everywhere in this flow.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -271,4 +278,61 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SAME;
```
Generic thread?

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java:
```diff
@@ -271,4 +278,61 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+
+                @Override
+                public void handleException(TransportException exp) {
+                    logger.warn("Failure occurred while dumping connection for decommission nodes - ", exp.unwrapCause());
```
logger.error?

imRishN replied:
logger.error(msg,e)

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -305,10 +308,65 @@ public void onFailure(Exception e) {
+            public void onResponse(DecommissionStatus status) {
+                logger.info("updated the decommission status to [{}]", status);
+                // set the weights
+                checkWeightsAndScheduleDecommission(decommissionRequest);
```
schedule decommission only

pranikum replied:
Done. Changed the method name.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -305,10 +308,65 @@ public void onFailure(Exception e) {
+        }, timeoutForNodeDraining, ThreadPool.Names.GENERIC);
+    }
+
+    private void failDecommissionedNodes(ClusterState state, DecommissionAttribute decommissionAttribute) {
```
Send the decommissioned nodes from the above set only

pranikum replied:
Done.

Bukhtawar commented on server/src/internalClusterTest/java/org/opensearch/cluster/coordination/AwarenessAttributeDecommissionIT.java:
```diff
@@ -104,38 +104,52 @@ public void testDecommissionStatusUpdatePublishedToAllNodes() throws ExecutionEx
         DecommissionResponse decommissionResponse = client().execute(DecommissionAction.INSTANCE, decommissionRequest).get();
         assertTrue(decommissionResponse.isAcknowledged());
 
+        logger.info("--> Recieved decommissioning nodes in zone {}", 'c');
+        // Keep some delay for scheduler to invoke decommission flow
+        Thread.sleep(3000);
```
nit: typo
3s sleep will slow down tests significantly can we try lower?

pranikum replied:
Reduced to 100 milis. Tried few times. Can see it passing all the times. Looks safe.

Bukhtawar commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestDecommissionAction.java:
```diff
@@ -49,6 +49,9 @@ DecommissionRequest createRequest(RestRequest request) throws IOException {
         String attributeValue = request.param("awareness_attribute_value");
-        return decommissionRequest.setDecommissionAttribute(new DecommissionAttribute(attributeName, attributeValue));
+        decommissionRequest.setDecommissionAttribute(new DecommissionAttribute(attributeName, attributeValue));
+        return decommissionRequest.setDrainingTimeout(
+            request.paramAsTime("draining_timeout", DecommissionRequest.DEFAULT_NODE_DRAINING_TIMEOUT)
```
Since the system isn't doing an explicit draining, should we call this as `delay_timeout` and also set a query param that says ?no_delay=true which sets the timeout at 0s

pranikum replied:
Done.

Bukhtawar commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -309,17 +312,72 @@ public void onFailure(Exception e) {
+
+        Set<DiscoveryNode> decommissionedNodes = filterNodesWithDecommissionAttribute(state, decommissionAttribute, false);
+        // Wait for timeout to happen. Log the active connection before decommissioning of nodes.
+        transportService.getThreadPool().schedule(() -> {
```
Check if the delay is zero if it is it doesn't make sense to fork and we can continue in the same thread?

Bukhtawar commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -309,17 +312,72 @@ public void onFailure(Exception e) {
     }
 
-    private void failDecommissionedNodes(ClusterState state) {
+    private void drainNodesWithDecommissionedAttribute(DecommissionRequest decommissionRequest) {
```
Can we skip draining if `no_delay` (timeout=0) flag is set and avoid a cluster state update altogether?

pranikum replied:
Added no_delay param for skipping skipping the delay.

Bukhtawar replied:
No can we avoid the `DRAINING` cluster state update altogether and go to `INPROGRESS`

pranikum replied:
I have created a task for the State removal. We have kept it for better debugging in case of decommission request failure

https://github.com/opensearch-project/OpenSearch/issues/4866

pranikum replied:
We have removed Draining state when no_Delay is set to true.

Bukhtawar commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -309,17 +312,72 @@ public void onFailure(Exception e) {
+
+    void checkDecommissionAttributeAndScheduleDecommission(DecommissionRequest decommissionRequest) {
+        ClusterState state = clusterService.getClusterApplierService().state();
+
```
nit: avoid unnecessary line breaks

Bukhtawar commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -309,17 +312,72 @@ public void onFailure(Exception e) {
+        ClusterState state = clusterService.getClusterApplierService().state();
+
+        DecommissionAttributeMetadata decommissionAttributeMetadata = state.metadata().decommissionAttributeMetadata();
+        if (decommissionAttributeMetadata == null) {
+            clearVotingConfigExclusionAndUpdateStatus(false, false);
+            return;
+        }
```
if metadata is null why should voting config not be reset back? I am assuming this is about the race with recommission? So why would recommission not reset voting config?

imRishN replied:
Recommission does that. We can remove this

pranikum replied:
Done.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -309,17 +312,72 @@ public void onFailure(Exception e) {
+        });
+    }
+
+    void checkDecommissionAttributeAndScheduleDecommission(DecommissionRequest decommissionRequest) {
+        ClusterState state = clusterService.getClusterApplierService().state();
+
+        DecommissionAttributeMetadata decommissionAttributeMetadata = state.metadata().decommissionAttributeMetadata();
+        if (decommissionAttributeMetadata == null) {
+            clearVotingConfigExclusionAndUpdateStatus(false, false);
+            return;
+        }
+        assert decommissionAttributeMetadata.status().equals(DecommissionStatus.DRAINING)
+            : "Unexpected status encountered while decommissioning nodes.";
+
+        // Schedule the node decommission process after the weights are successfully set.
+        scheduleNodesDecommissionOnTimeout(state, decommissionRequest.getDrainingTimeout());
+    }
```
we can avoid this method entirely

pranikum replied:
Keeping it to decide on schedule vs plain call in same thread.

imRishN commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -309,17 +312,72 @@ public void onFailure(Exception e) {
+    void scheduleNodesDecommissionOnTimeout(ClusterState state, TimeValue timeoutForNodeDraining) {
         // this method ensures no matter what, we always exit from this function after clearing the voting config exclusion
         DecommissionAttributeMetadata decommissionAttributeMetadata = state.metadata().decommissionAttributeMetadata();
         DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
```
check for attribute is being null here

pranikum replied:
Keeping the previous method.

pranikum commented on server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java:
```diff
@@ -309,17 +312,72 @@ public void onFailure(Exception e) {
+        }, timeoutForNodeDraining, ThreadPool.Names.GENERIC);
+    }
+
+    private void failDecommissionedNodes(Set<DiscoveryNode> decommissionedNodes, DecommissionAttribute decommissionAttribute) {
```
Used in logging .. keeping it

Bukhtawar commented on server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java:
```diff
@@ -90,8 +94,16 @@ public DecommissionAttribute getDecommissionAttribute() {
+    public void setNoDelay(boolean noDelay) {
+        this.noDelay = noDelay;
+    }
+
```
we need to guarantee only one of no_delay and delay timeout is set

pranikum replied:
We have done that in DecommissionRequest.

Bukhtawar commented on server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java:
```diff
@@ -58,13 +76,36 @@ public DecommissionRequest setDecommissionAttribute(DecommissionAttribute decomm
+        return this.delayTimeout;
+    }
+
+    public void setNoDelay(boolean noDelay) {
+        this.noDelay = noDelay;
+    }
+
+    public boolean isNoDelay() {
+        return noDelay;
+    }
```
We can simplify this by keeping only delayTimeout to zero. No delay was a flag for better API experience

pranikum replied:
Have removed the param from `[RestDecommissionAction.java](https://github.com/opensearch-project/OpenSearch/pull/4586/files#diff-65a6fad1a8dab03e81118e6dc2e25f256a4abc3a26dd9004e45709c9db3a1992)`

Bukhtawar reviewed (approved):
Lets add assertions in the integration tests for voting configuration values

imRishN commented on server/src/main/java/org/opensearch/action/admin/cluster/decommission/awareness/put/DecommissionRequest.java:
```diff
@@ -39,12 +47,14 @@ public DecommissionRequest(DecommissionAttribute decommissionAttribute) {
     public DecommissionRequest(StreamInput in) throws IOException {
         super(in);
         decommissionAttribute = new DecommissionAttribute(in);
+        this.delayTimeout = in.readTimeValue();
```
We missed adding `noDelay` here. With this no delay would never be read from the request
